<html>
  <head>
    <title>Tutorial</title>
  </head>

<body>

<h1>Ontologizer Tutorial</h1>

<h2>Setting Up the Ontologizer</h2>

<ol>

<li>
Download the file <A HREF="http://compbio.charite.de/tl_files/ontologizer/examples/yeastSampleFiles.zip">http://compbio.charite.de/tl_files/ontologizer/examples/yeastSampleFiles.zip</A> and unpack it into a directory of your choice, e.g,
the desktop. The archive contains sets of genes up- or downregulated following treatment with sulfometuron methyl, which is an inhibitor of amino acid biosynthesis. Data
was gathered from <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=11015603&itool=iconfft&query_hl=1&itool=pubmed_docsum">Jia et al. (2000) Global expression profiling of yeast treated with an inhibitor of amino acid biosynthesis, sulfometuron methyl</A>.
</li>

<li>
If you need a proxy to access the internet, please open the <i>Preferences Window</i> via the <i>Window > Preferences...</i> menu entry within Ontologizer. 
<br /><img src="images/menu-preferences.png" />
</li>

<li>
Enter your proxy configuration in the appropriate line and then press <i>Ok</i>.
<br /><img src="images/preferences.png" />
</li>
</ol>

<h2>Downloading the Files for the Tutorial Project</h2>
<ol>
<li>For GO analysis, two sources of information (files) are required. 
<ul>
<li>The <em>gene_ontology_edit.obo</em> file contains definitions of individual GO terms
such as <em>Development</em> or <em>DNA helicase</em>. We will refer to this file as the <strong>definition file</strong>.</li>
<li>A species-specific <strong>association file</strong>. Such files have been produced for a number of organisms and are listed at the GO homepage http://www.geneontology.org. For example, the association file for yeast (Saccharomyces Genome Database) is <em>gene_association.sgd</em>.</li>
</ul>
<li>
The Ontologizer can automatically download the definition file (gene_ontology_edit.obo) and several predefined association files (e.g. for
 human, mouse, fruit fly, yeast, and PDB) from the web in the background. As we will see, the Ontologizer downloads these files automatically from the web, 
 but if the user desires, files can also be downloaded by hand and loaded into the Ontologizer using a standard file selection dialog.</li>
<li>The GO website (http://www.geneontology.org) contains downloads for the GO definition file as well as association files for numerous other organisms.</li>
<li>Before starting this tutorial, we recommend that you download the definition file and the association file
using the automatic download wizard in order to check your internet and (if applicable) proxy configurations. Open the <i>Work Set Window</i> by using the
<i>Window > File Sets...</i> menu entry.
</li>

<li>
Within the tree, look for an item called <i>Yeast</i>. Click on the <i>Definitions</i> entry within this subtree and then the <i>Download Now</i> button. Repeat this with the
<i>Association</i> entry. Both files should be downloaded now. If an error occurs, check your internet connection and verify the proxy settings. On success the second column should
change to "Cached". <br />
<img src="images/filesets.png" />
</li>
</ol>


<h2>File Sets</h2>
<ul><LI>Each File Set contains a definition file and an association file. Each project uses a File Set to perform the analysis. This is done so that users 
can manage different file configurations easily, allowing for instance different versions of the definition file to be used for different projects, 
as it may be useful to use the same version of the definition file (which is frequently updated at the GO website) for development.</LI></ul>



<h2>User-Supplied Files</h2>
<ul><LI>In order to analyze your experimental data, you need to prepare one file for each of the groups of interest in your experiment (For instance, this might be a list of genes differentially expressed at different time points). We refer to such groups as <strong>study sets</strong>. Additionally, you need to indicate the <strong>population set</strong>. In general, this will be a list of all genes that were test (for instance, all genes represented on a microarray). The genes should
be listed one on a line in plain text (Alternatively, FASTA files can be used if desired if the name of the gene directly follows the '>' sign).
</LI>
<li>For this tutorial, you can download the yeast study and populations files from the Ontologizer website: http://www.charite.de/ch/medgen/ontologizer/howto/index.html. Unpack these files before use.</li>
</ul>



<h2>Creating a New Project</h2>
<p>
<ol>
 <li>In order to create the new project, press the <i>New Project</i> button within the toolbar of Ontologizer's main window or select
     the <i>Project > New > Project...</i> menu entry.
     <br /><img src="images/new-project.png" />
 </li>
 <li> This brings up   the <strong>New Project Wizard</strong>. First enter a name for the project. For the tutorial, 
 enter <i>suflometuronMethyl</i> into the <i>Project Name</i> textfield then press the <i>Next</i> button  to proceed to the next page.
      
      <br /><img src="images/name-of-the-project.png"/> </li>
 <li>
	Here you need to indicate the
	<em>definition file</em> (via the <i>Ontology</i> text field) and the <em>association file</em>  (via
	the <i>Association</i> text field). The Ontologizer comes with predefined File Sets for frequently
	used species that can be automatically downloaded. We have downloaded the File Set for Yeast above. If we hadn't, the Ontologizer would
	now automatically download these files in the background. For this tutorial, click on the <i>File Set</i> combo box and choose <i>Yeast</i>. Then press
	<i>Next</i> which brings you to the <strong>Population Edit</strong> page.
	<br /><img src="images/new-project-fileset.png" />
 </li>
 
 <li>
	Now enter the genes of the <strong>population set</strong>. Use the study set/population set example files downloaded from 
	the Ontologizer homepage as described above.
	Drag &#038; Drop the file called
	<i>population.txt</i>  into the <strong>gene editor</strong> field or use a File Selection Dialog by clicking on 
	the <em>Append Set...</em> button. Notice that names of genes with GO annotations are highlighted (you may have to wait 
	for completion of downloads or parsing before seeing highlighting). You can hover the mouse over these entries to see more 
	information about the gene's annotation. Proceed by clicking on the <i>Next</i>
	button.
	<br /><img src="images/population.png" /> 
 </li>

 <li>
  Drag &#038; Drop a study file into editor area (again, alternatively, you can use a
 file selection dialog by clicking on the <em>Append Set...</em> button). Press <i>Next</i> and repeat the procedure for each study set (file).

  <br /><img src="images/study.png" /> 
 </li>
 
 <li>
  Press <i>Finish</i> when you added the last study set. The <strong>New Project Wizard</strong> window closes and
  you should now see your new project <i>suflometuronMethyl</i> appearing in the main
  window.
 </li>
</ol>

<h2>Performing the Analysis</h2>
<p>The Ontologizer offers multiple methods for searching for GO term overrepresentation and for multiple testing correction. For more information
on these topics please consult the Ontologizer homepage, where you will also find links to publications describing the Ontologizer. For the purposes
of this tutorial, we will use the Parent-Child Union Methods with a Bonferroni multiple testing correction. </p>
<ol>
 <li>Within the main window, select our project which is <i>sulfometuronMethyl</i>.
     <br /><img src="images/main-with-project.png" />
 
 </li>
 <li>From the combo boxes in the tool bar, choose a <A HREF="3_howto.html#stats">calculation method</A> (first combo box), <i>Parent-Child-Union</i> and 
     and  the <i>Bonferroni</i> a <A HREF="3_howto.html#stats">multiple test correction</A> (second combo box). Then press <i>Analyze</i>.
  </li>
</ol>

<h2>Exploring the Results</h2>

<ol>
<li>
     The <strong>Results Window</strong> now appears. Depending on the size and number of the study sets and the type of multiple testing correction desired, the
analyis should complete in a few seconds to a few minutes. As individual study sets are completed, new tabs appear with the results. If you have used
all the files of this tutorial, you should see seven tab folders corresponding to the name of the study sets
     once analysis is completed. The first study set is activated and within the tab folder the results are presented in form of a table. 
     <br /><img src="images/result-initial.png" />
</li>

<li>
 Notice that the background of terms whose adjusted p-value falls below the significance level
 (as given by widget below the table) is colorized according to the sub-ontology and the rank.  
</li>

<li>
 Now click on one of the terms, e.g., <i>amino acid and derivative</i>. This refreshes the browser of the bottom part in the window to contain
 information about the term including the parents (more general terms), children (more specific terms) or the names of the genes, to which the term is
 annotated to.
</li>

<li>
 To get a graphical overview, press the <i>Preview Graph</i> button (the third from left in the toolbar). 
</li>

<li>
The graphs consists of all active terms as defined by
 the little checkboxes before every time, which by default are all signifcant terms. 
</li>

<li>
The parameter of the graph view (i.e., zoom factor, which extend is displayed) can be altered the button or context menu commands.
</li>

<li>

</li>

</ol>


</body>
</html>